src/SSymmetricCompact.jl

```diff
@@ -0,0 +1,239 @@
+    if 2 * L != N * (N + 1)
+        return :(throw(ArgumentError("Size mismatch in SSymmetricCompact parameters. Got dimension $N and length $L.")))
+    end
+    :(nothing)
+end
+
+Base.@pure triangularnumber(N::Int) = div(N * (N + 1), 2)
+Base.@pure triangularroot(L::Int) = div(isqrt(8 * L + 1) - 1, 2) # from quadratic formula
+
+lowertriangletype(::Type{SSymmetricCompact{N, T, L}}) where {N, T, L} = SVector{L, T}
+lowertriangletype(::Type{SSymmetricCompact{N, T}}) where {N, T} = SVector{triangularnumber(N), T}
+lowertriangletype(::Type{SSymmetricCompact{N}}) where {N} = SVector{triangularnumber(N)}
+
+@inline (::Type{SSymmetricCompact{N, T}})(lowertriangle::SVector{L}) where {N, T, L} = SSymmetricCompact{N, T, L}(lowertriangle)
+@inline (::Type{SSymmetricCompact{N}})(lowertriangle::SVector{L, T}) where {N, T, L} = SSymmetricCompact{N, T, L}(lowertriangle)
+
+@inline function SSymmetricCompact(lowertriangle::SVector{L, T}) where {T, L}
+    N = triangularroot(L)
+    SSymmetricCompact{N, T, L}(lowertriangle)
+end
+
+@generated function (::Type{SSymmetricCompact{N, T, L}})(a::Tuple) where {N, T, L}
+    expr = Vector{Expr}(undef, L)
+    i = 0
+    for col = 1 : N, row = col : N
+        index = N * (col - 1) + row
+        expr[i += 1] = :(a[$index])
+    end
+    quote
+        @_inline_meta
+        @inbounds return SSymmetricCompact{N, T, L}(SVector{L, T}(tuple($(expr...))))
+    end
+end
+
+@inline function (::Type{SSymmetricCompact{N, T}})(a::Tuple) where {N, T}
+    L = triangularnumber(N)
+    SSymmetricCompact{N, T, L}(a)
+end
+
+@inline (::Type{SSymmetricCompact{N}})(a::Tuple) where {N} = SSymmetricCompact{N, promote_tuple_eltype(a)}(a)
+@inline (::Type{SSymmetricCompact{N}})(a::NTuple{M, T}) where {N, T, M} = SSymmetricCompact{N, T}(a)
+@inline SSymmetricCompact(a::StaticMatrix{N, N, T}) where {N, T} = SSymmetricCompact{N, T}(a)
+
+@inline (::Type{SSC})(a::SSymmetricCompact) where {SSC <: SSymmetricCompact} = SSC(a.lowertriangle)
+@inline (::Type{SSC})(a::SSC) where {SSC <: SSymmetricCompact} = a
+
+@inline (::Type{SSC})(a::AbstractVector) where {SSC <: SSymmetricCompact} = SSC(convert(lowertriangletype(SSC), a))
+
+# TODO: similar_type overload?
+
+@generated function _symmetric_compact_indices(::Val{N}) where N
+    # Returns a Tuple{Pair{Int, Bool}} I such that for linear index i,
+    # * I[i][1] is the index into the lowertriangle field of an SSymmetricCompact{N};
+    # * I[i][2] is true iff i is an index into the lower triangle of an N × N matrix.
+    indexmat = Matrix{Pair{Int, Bool}}(undef, N, N)
+    i = 0
+    for col = 1 : N, row = 1 : N
+        indexmat[row, col] = if row >= col
+            (i += 1) => true
+        else
+            indexmat[col, row][1] => false
+        end
+    end
+    quote
+        Base.@_inline_meta
+        return $(tuple(indexmat...))
+    end
+end
+
+Base.@propagate_inbounds function Base.getindex(a::SSymmetricCompact{N}, i::Int) where {N}
```

Comment:
Why implement this rather than `getindex(a, i::Int, j::Int)` and make it `IndexCartesian`?

Reply:
The main reasons are that `getindex(a, i::Int, j::Int)` would also not be trivial, requiring a solution similar to `_symmetric_compact_indices` to be fast, and a lot of functionality inside of StaticArrays implicitly assumes that all `StaticArray`s are `IndexLinear`, e.g.:

https://github.com/JuliaArrays/StaticArrays.jl/blob/407c65fb2ae652a439d09d82edf1a22eaab58dc6/src/mapreduce.jl#L34-L35

Reply:
Regarding the nontrivial, doesn't
```julia
i, j = i >= j ? (i, j) : (j, i)
idx = i + (j-1)*N
```
do the job?

If you've not benchmarked it or checked generated code, this seems to be worth doing. But most likely you have and you know that it's all fine.

Reply:
No, that's not quite right.

```julia
julia> N = 3; L = StaticArrays.triangularnumber(N)
6

julia> m = SSymmetricCompact(SVector{L}(1 : L))
3×3 SSymmetricCompact{3,Int64,6}:
 1  2  3
 2  4  5
 3  5  6

julia> i, j = 2, 3;

julia> m[i, j]
5

julia> i, j = i >= j ? (i, j) : (j, i)
(3, 2)

julia> idx = i + (j-1)*N
6

julia> m.lowertriangle[idx]
6
```

Reply:
D'oh!  Try
```julia
idx = i - j + 1 + ((j-1)*(2N+2-j)÷2)
```
The point being: your implementation *looks* slow, though perhaps it's not.

Reply:
That's not quite right either I'm afraid:

```julia
```

As I showed in https://github.com/JuliaArrays/StaticArrays.jl/pull/358#issuecomment-488432426, the compiler does a very good job on the whole `_symmetric_compact_indices` thing. Basically, it computes that tuple at compile time for a given `N`, and then the `getindex` method simply indexes into that constant tuple.

Reply:
Oh wait, I should still add the `i, j = i >= j ? (i, j) : (j, i)`; it works then. I'll benchmark.

Reply:
Alright, with

```julia
Base.IndexStyle(::Type{<:SSymmetricCompact}) = Base.IndexCartesian()

Base.@propagate_inbounds function Base.getindex(a::SSymmetricCompact{N}, i::Int) where N
    # override StaticArray overload, replace with AbstractArray version.
    # Base._getindex(IndexStyle(a), a, Base.to_indices(a, (i,))...)
    c, r = divrem(i - 1, N)
    c, r = c + 1, r + 1
    a[r, c]
end

Base.@propagate_inbounds function Base.getindex(a::SSymmetricCompact{N}, i::Int, j::Int) where N
    i, j = i >= j ? (i, j) : (j, i)
    idx = i - j + 1 + ((j-1)*(2N+2-j)÷2)
    x = a.lowertriangle[idx]
end
```

we get

```julia
julia> @btime a[i] setup = begin
           a = SSymmetricCompact(rand(SMatrix{6, 6}))
           i = rand(1 : length(a))
       end
  3.272 ns (0 allocations: 0 bytes)
```

whereas before,

```julia
julia> @btime a[i] setup = begin
           a = SSymmetricCompact(rand(SMatrix{6, 6}))
           i = rand(1 : length(a))
       end
  1.783 ns (0 allocations: 0 bytes)
```

Do note that multiplying two `SSymmetricCompact`s results in the same runtime before and after this change, as the indices are known at compile time and so it looks like the compiler is able to constant-propagate everything in either case. But just randomly linearly indexing into an `SSymmetricCompact` is faster with the `_symmetric_compact_indices` approach.

Reply:
Right, my proposal is really for `IndexCartesian`, where you never have to call `divrem` (which is insanely slow).

But I just benchmarked with
```julia
Base.@propagate_inbounds function Base.getindex(a::SSymmetricCompact{N}, i::Int, j::Int) where {N}
    i, j, lower = j > i ? (j, i, false) : (i, j, true)
    idx = i - j + 1 + ((j-1)*(2N+2-j))>>UInt(1)
    @inbounds value = a.lowertriangle[idx]
    return lower ? value : transpose(value)
end
```
and to my surprise it's no better. So I'm happy going with yours.

Reply:
For constant `y` the compiler does a very good job on `div(x, y)` and `divrem(x, y)`, almost to the point where it produces the same code as the manual `>> UInt(1)` optimization:

```julia
julia> f(x) = div(x, 2);

julia> g(x) = x >> UInt(1);

julia> @code_native f(3)
        .text
; ┌ @ REPL[15]:1 within `f'
; │┌ @ REPL[15]:1 within `div'
        movq    %rdi, %rax
        shrq    $63, %rax
        leaq    (%rax,%rdi), %rax
        sarq    %rax
; │└
        retq
        nop
; └

julia> @code_native g(3)
        .text
; ┌ @ REPL[16]:1 within `g'
; │┌ @ REPL[16]:1 within `>>'
        sarq    %rdi
; │└
        movq    %rdi, %rax
        retq
        nopw    (%rax,%rax)
; └
```

Similarly for `divrem(i, N)` for fixed `N`.